### PR TITLE
Don't drop mojo callback in LocalStorageImpl::DeleteStorage.

### PR DESCRIPTION
**Pull request by goodov**
Uplift of #9629
Resolves https://github.com/brave/brave-browser/issues/17325 (the original issue on browser shutdown which is already merged to ```master``` and ```1.30.x```)
Resolves https://github.com/brave/brave-browser/issues/17859 (the production issue which breaks ```localStorage``` in some cases, but is also fixed by this change)

Pre-approval checklist: 
- [x] You have tested your change on Nightly. 
- [ ] This contains text which needs to be translated. 
    - [ ] There are more than 7 days before the release. 
    - [ ] I've notified folks in #l10n on Slack that translations are needed. 
- [x] The PR milestones match the branch they are landing to. 

Pre-merge checklist: 
- [x] You have checked CI and the builds, lint, and tests all pass or are not related to your PR. 

Post-merge checklist: 
- [x] The associated issue milestone is set to the smallest version that the changes is landed on.

### CODE REVIEW

**Comment by goodov** [comment]
This change already exists in ```master``` and ```1.30.x```.

**Review by kjozwiak — APPROVED** [review]
Uplift into `1.29.x` approved after deliberating with @brave/uplift-approvers. QA has verified the PR on Nightly as per https://github.com/brave/brave-core/pull/9629#issuecomment-914928888.

**Pull request merged**
